### ISSUE TITLE
Broken from_iter_instead_of_collect / FromIterator::from_iter lint

### ISSUE THREAD

**Issue by caseykneale** [rust-lang/rust-clippy#10104]
### Summary

I think the code examples explain it better than I can.

### Reproducer

I tried this code:

```rust
        let _break_clippy: HashSet<String> = HashSet::from_iter(
            vec![vec!["ab", "cd"],vec!["xy", "za"]]
                .iter()
                .map(|x| (*x).iter().fold(String::new(),|a,b| a+b)),
        );
```

I expected to see this happen: Honestly I expect clippy to either be okay with this or suggest valid rust syntax. Also this isn't the actual code that is failing the lint, but it is a MWE so don't judge too harshly :).

Instead, this happened:
```rust
error: usage of `FromIterator::from_iter`
    |
262 |           let _break_clippy: HashSet<String> = HashSet::from_iter(
    |  ______________________________________________^
263 | |             vec![vec!["ab", "cd"],vec!["xy", "za"]]
264 | |                 .iter()
265 | |                 .map(|x| (*x).iter().fold(String::new(),|a,b| a+b)),
266 | |         );
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#from_iter_instead_of_collect
help: use `.collect()` instead of `::from_iter()`
    |
262 ~         let _break_clippy: HashSet<String> = vec![vec!["ab", "cd"],vec!["xy", "za"]]
263 +                 .iter()
264 +                 .map(|x| (*x).iter().fold(String::new(),|a,b| a+b)).collect::<HashSet::from_iter(
265 +             vec![vec!["ab", "cd"],vec!["xy", "za"]]
266 +                 .iter()
267 ~                 .map(|x| (*x).iter().fold(String<_>>();
```

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-apple-darwin
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by feniljain** [comment]
Few things which comes to my mind:
- Lint suggests to use `collect` instead of `from_iter`, and it is named `from_iter_instead_of_collect`, seems weird to me. Maybe I am just missing something?
- I tried to understand the reasoning behind the lint, you can understand more using thread here: #6550 
- Keeping this in mind, I think that lint being suggested here makes sense?

**Comment by Manishearth** [comment]
> * Lint suggests to use `collect` instead of `from_iter`, and it is named `from_iter_instead_of_collect`, seems weird to me. Maybe I am just missing something?

Lint names are named according to the code pattern they wish to forbid, not the fix they apply.
